File: tp_factorielle/c/bench_fact.c
// c/bench_fact.c - À compléter
#include <stdio.h>
#include <time.h>
#include <stdbool.h>
#include "fact.h"

double bench(unsigned int n, int iters) {
    // TODO-1: chronométrer iters appels à fact_ull(n) en ignorant la valeur
    clock_t start = clock();
    bool overflow;
    for (int i = 0; i < iters; i++) {
        fact_ull(n, &overflow);
    }
    clock_t end = clock();
    return ((double)(end - start)) / CLOCKS_PER_SEC;
}

int main(void) {
    // TODO-2: appeler bench pour n=12 et n=20, iters=100000; afficher les durées
    printf("=== Benchmarks Factorielle C ===\n\n");
    
    int iters = 100000;
    unsigned int test_values[] = {12, 20};
    int num_tests = sizeof(test_values) / sizeof(test_values[0]);
    
    for (int i = 0; i < num_tests; i++) {
        unsigned int n = test_values[i];
        double duration = bench(n, iters);
        printf("fact_ull(%u) - %.6fs pour %d appels\n", n, duration, iters);
        printf("  Moyenne: %.2f μs par appel\n", duration / iters * 1000000);
    }
    
    return 0;
}

File: tp_factorielle/python/bench_fact.py
# python/bench_fact.py - À compléter
from time import perf_counter
from fact import fact

def bench(fn, n, iters=100_000):
    """
    Retourne la durée (en secondes) pour iters appels de fn(n).
    Attention: mesures indicatives (bruit, cache, turbo CPU...).
    """
    # TODO: chronométrer une boucle d'appels
    start = perf_counter()
    for _ in range(iters):
        fn(n)
    end = perf_counter()
    return end - start

def main():
    # TODO: pour n in (12, 20), appeler bench et afficher un résumé
    print("=== Benchmarks Factorielle ===\n")
    
    for n in [12, 20]:
        duration = bench(fact, n, 100_000)
        print(f"fact({n}) - {duration:.6f}s pour 100,000 appels")
        print(f"  Moyenne: {duration/100_000*1_000_000:.2f} μs par appel")

if __name__ == "__main__":
    main()